### PR TITLE
Move channels out of listpeers, issue #4729

### PR DESCRIPTION
**Pull request by rl-d**
See issue for details.

### CODE REVIEW

File: lightningd/peer_control.c [hunk 1]
```diff
@@ -1611,6 +1613,75 @@ static const struct json_command listpeers_command = {
 /* Comment added to satisfice AUTODATA */
 AUTODATA(json_command, &listpeers_command);
 
+
+static void json_add_peerchannels(struct lightningd *ld,
+                          struct json_stream *response,
+                          struct peer *peer,
+			  enum channel_state *state,
```
Comment (vincenzopalazzo) [inline lightningd/peer_control.c:1620]: Here there is a missing space.

File: lightningd/peer_control.c [hunk 2]
```diff
@@ -1611,6 +1613,75 @@ static const struct json_command listpeers_command = {
...
+{
+	enum log_level *ll;
+	struct node_id *peer_id;
+	enum channel_state *state;
+	struct peer *peer;
+	struct json_stream *response;
+
+	if (!param(cmd, buffer, params,
+		   p_opt("id", param_node_id, &peer_id),
+		   p_opt("state", param_channel_state, &state),
```
Comment (vincenzopalazzo) [inline lightningd/peer_control.c:1654]: I like to maintain the same convention for all the parameters, when we filter by some status, we use the word "status" instead of "state". To avoid confusion for the user.

File: lightningd/peer_control.c [hunk 3]
```diff
@@ -1611,6 +1613,75 @@ static const struct json_command listpeers_command = {
...
+static void json_add_peerchannels(struct lightningd *ld,
+                          struct json_stream *response,
+                          struct peer *peer,
+			  enum channel_state *state,
+                          const enum log_level *ll)
+{
+	struct channel *channel;
+
+        if (!state)
+	        json_add_uncommitted_channel(response, peer->uncommitted_channel);
```
Comment (vincenzopalazzo) [inline lightningd/peer_control.c:1626]: I'm sure that I'm missing some things here, but I didn't catch this part here of your code. 

Why if the state is unspecified in the command you are returning only the uncommitted channels? The state here is useful to get the channel with an ON_CHAIN (closed state). 

I think it is rare to filter the channels by a status here like "CHANNELD_NORMAL", I usually preferer with API to simplify the life of the user. I would be happy to have only a couple of status propriety like "closed", "normal", "pending". So in this way, you can have that the following state correspond to the 

- closed -> (ONCHAIN, CLOSED)
- normal -> (CHANNELD_NORMAL)
- pending -> (CHANNELD_SHUTTING_DOWN, CLOSINGD_SIGEXCHANGE, CLOSINGD_COMPLETE)

I think this makes cooler the API of the status, now I know that CLOSINGD_SIGEXCHANGE can be a pending status of the channel, this means that some operations need to happen.

p.s: I think this is a good moment to query the DB to retrieval the closed channels with the status "closed"

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
Some comments on this PR.

In addition, you seem that you are overriding an old lnprototest dependencies. Usually, this is done by a rebase (or merge).

In the commit https://github.com/ElementsProject/lightning/pull/4770/commits/4572dfaae985b5e7b97ef6b6892eea3f3f1ef329 you need to put the change-log in the body of the commit (usually as the last thing, or before your signature)

I will make a summary, of your work on Rusty TODO list

 - [X] Create a listpeerchannels command next to the listpeers implementation, which returns a "channels" array.
- [X] Write the doc/schemas/listpeerchannels.schema.json file (copy existing ones)
 - [ ] Write some tests (steal from existing listpeers tests), put Changelog-Added: JSON-RPC: listpeerchannels new command.
 If deprecated_apis is false, don't include "channels" in listpeers.
 - [ ] Update listpeers documentation to refer to listpeerchannels and remove channels[] from doc/schemas/listpeers.schema.json
 - [X] Make sure to put a Changelog-Deprecated: in the commit msg for the commit which removes from listpeers.
 - [X] Celebrate!

P.S: The big missing here is the unit testing on the new command and the old command with the deprecated API. I will fully test this PR locally later in my day (the unit testing is another proof before I ran the code that your code is runnable).

**Comment by rl-d** [comment]
Thank you @vincenzopalazzo for your review. I want to do the unit testing now, but need to ask somebody some newbie questions... I've written you in Telegram.

File: lightningd/peer_control.c (hunk 3 above)
Comment (rl-d) [reply lightningd/peer_control.c:1626]: Ok, please allow me to explain the case with unspecified state. The original listpeers code did not have a state filter, and it returned uncommitted channels AND all the other relevant channels, and I simply retained that logic here, with the exception that if the state filter is specified, then uncommitted channels are omitted from the result. To be clear, if no state is specified, the result will contain uncommitted channels AND all other channels just like the original behavior of listpeers.

I think the DB retrieval of closed channels should be kept separate from this issue, to be handled by the plugin I am working on. Perhaps not to muddy the waters, we should even take out the filter option from this PR altogether.

**Comment by vincenzopalazzo** [comment]
Ah right, we have this issue #4773 with the make check command, I will try to find a solution later in this day. However, as suggested already on telegram, you can work around with the following command `DEVELOPER=1 EXPERIMENTAL_DUAL_FUND=0 python3 -m pytest tests/test_pay.py::test_delete_inactive_offers_two`

I will jump in your comment with a fresh mind later, thank

**Comment by rl-d** [comment]
@vincenzopalazzo, in your summary, you did not mark that I completed the item "Update listpeers documentation to refer to listpeerchannels and remove channels[] from doc/schemas/listpeers.schema.json". I believe maybe you just overlooked that the documentation for listpeers has been updated. Or did I miss something with that?

File: lightningd/peer_control.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (rl-d) [reply lightningd/peer_control.c]: I have made the change and updated the documentation accordingly.

File: lightningd/peer_control.c (hunk 3 above)
Comment (vincenzopalazzo) [reply lightningd/peer_control.c:1626]: >Ok, please allow me to explain the case with unspecified state. The original listpeers code did not have a state filter, and it returned uncommitted channels AND all the other relevant channels, and I simply retained that logic here, with the exception that if the state filter is specified, then uncommitted channels are omitted from the result. To be clear, if no state is specified, the result will contain uncommitted channels AND all other channels just like the original behavior of listpeers.

Usually in the c-lightning commands when you don't specify the status the command line returns all the data without filter by status. If the status is specified, there is filtering with the status applied to it.

>I think the DB retrieval of closed channels should be kept separate from this issue, to be handled by the plugin I am working on. Perhaps not to muddy the waters, we should even take out the filter option from this PR altogether.

I think that if you specify the status here, you need to add also the query to the closed channels, otherwise, if you query by CLOSED status you have no result, but this in a node that works for a long time can be wrong (and also confusing for the end user?).

**Comment by rl-d** [comment]
@vincenzopalazzo I've tried to investigate these two things you mentioned:

> In addition, you seem that you are overriding an old lnprototest dependencies. Usually, this is done by a rebase (or merge).
> 
> In the commit 4572dfa you need to put the change-log in the body of the commit (usually as the last thing, or before your signature)

And I am reluctant to dabble with frowned-upon force pushes and frowned-upon rebase operations. If the tools (git/hub) are such that even a simple code change or editing a commit message requires risky interventions for a github beginner who is contributing his first trivial code change, I think maybe someone with more github experience should tackle that or at least help me learn what I did wrong.

**Comment by rl-d** [comment]
What I could do is copy all the changes into a fresh clone of the repo and commit them with a new message. That should solve both of those problems... I'll try that.

**Comment by vincenzopalazzo** [comment]
As usual, when we will start to learn some things there are always some things frustrating, in this case, it is git, and I'm sorry for that. I'm not a big expert, and usually, I'm very bad to suggest some things, but I will try.

- lnprototest is pulled with `git submodule` and as I noted you have an old version of it in the git history, when you get the new version of the lightning master with `git pull`, you can see that the version is different with the `git log` inside the lnprototest dir. [See how you can update a submodule](https://stackoverflow.com/a/5828396/14933807).
- The changelog is a rule of lightning repository https://lightning.readthedocs.io/STYLE.html#changelog-entries-in-commit-messages basically it helps the team to track the change in automating way, without breaking the mind around the git history. In your code commit, a `Changelog-Added:` it is missed, I usually put this message in the body of the commit (when you make git commit, git will open an editor to typing title and body, if you use git commit -m "No body here" you have only the possibility to type the title).

I hope that this doesn't bring confusion.

>What I could do is copy all the changes into a fresh clone of the repo and commit them with a new message. That should solve both of those problems... I'll try that.

This always works, but unfortunately, you need another PR.

**Comment by rl-d** [comment]
Ok, I don't really understand how you can tell from the git log vs. the git pull output that the submodule version is wrong...you're talking to a complete newbie when it comes to git. Nevertheless, I tried doing a git checkout in the lnprototest directory followed by a git pull to update it, then a top-level commit and subsequent build check and still got the same "fetchinvoice-noconnect documented but not in schema!" error. At this point, I'm not even sure if that was the error you were linking to the submodule version mismatch or not, but I also tried to run various unit tests and got more errors than success. It's kinda hard to write a test case when I can't even run the existing ones successfully. At this point, I'm afraid I'm not becoming very useful and my other work is starting to need my attention. I wanted to get involved and contribute something to this project, but I feel like I'm missing too much in the way of project orientation and git experience to be effective. Perhaps @vincenzopalazzo your offer to take assignment of this issue is truly in the best interest of the team.

**Comment by cdecker** [comment]
> Ok, I don't really understand how you can tell from the git log vs. the git pull output that the submodule version is wrong...you're talking to a complete newbie when it comes to git.

One way to tell is this entry in the Github interface: https://github.com/ElementsProject/lightning/pull/4770/files#diff-c17dbe2829c96cb4e940337269ba77c1c61a025d94743f563d9ccaf08ed2cc97
It basically just lists the files in the submodule that were updated. Submodules are one of the hardest part of `git` and I definitely don't fully understand how a practical workflow can be structured around them, we've just been winging it until now :wink: 

If you want I can try to clean up the branch, report the steps involved, and then force-push here. You'd then just have to `git pull -f` to overwrite your local copy based on my fixed up branch and we'd be in a good place again ^^

> Nevertheless, I tried doing a git checkout in the lnprototest directory followed by a git pull to update it, then a top-level commit and subsequent build check and still got the same "fetchinvoice-noconnect documented but not in schema!" error. At this point, I'm not even sure if that was the error you were linking to the submodule version mismatch or not, but I also tried to run various unit tests and got more errors than success.

If you want to sync up with the referenced submodule version, a `git submodule update` should do just that. It'll check out the version that is referenced in the main repository, and get you to a good state.

> It's kinda hard to write a test case when I can't even run the existing ones successfully. At this point, I'm afraid I'm not becoming very useful and my other work is starting to need my attention.

If I'm perfectly honest I just run `make pytest` which skips `lnprototest` altogether, which may or may not have some issues with some configuration options, and is sometimes flaky. `make pytest` will run solely the tests in the `tests` subdirectory which is also run in CI with a large variety of configurations, and should therefore always work. If you just want to selectively run a single test you can also run this:

```
pytest tests -vvv -k test_name
```

if you also add the `-s` flag you'll see the output fly by and see if things get stuck somewhere.

> I wanted to get involved and contribute something to this project, but I feel like I'm missing too much in the way of project orientation and git experience to be effective.

No worries, there is a lot of tribal knowledge that isn't written down, we try to formalize and create more resources over time, but we're hppy to help you get started. Don't worry if something doesn't work right away, just ping us and we can walk you through the steps :-)

> Perhaps @vincenzopalazzo your offer to take assignment of this issue is truly in the best interest of the team.

You're making awesome progress, don't let yourself be discouraged :-)

**Comment by rl-d** [comment]
> Submodules are one of the hardest part of `git` and I definitely don't fully understand how a practical workflow can be structured around them, we've just been winging it until now 😉

😄 It's making sense now...and I was able to resolve the submodule version mismatch. (I had overlooked some really obvious things...) Thanks for taking the time to nudge me once more in the right direction.

> > I wanted to get involved and contribute something to this project, but I feel like I'm missing too much in the way of project orientation and git experience to be effective.
> 
> No worries, there is a lot of tribal knowledge that isn't written down, we try to formalize and create more resources over time, but we're hppy to help you get started. Don't worry if something doesn't work right away, just ping us and we can walk you through the steps :-)
> 
> > Perhaps @vincenzopalazzo your offer to take assignment of this issue is truly in the best interest of the team.
> 
> You're making awesome progress, don't let yourself be discouraged :-)

Thanks for taking the time to respond with words of encouragement and a willingness to help (and I also appreciate the existing documentation insofar as it goes). Git got clearer to me today. The biggest problem I am still struggling with at this point is running the tests, which of course is critical to any development. For example, `make pytest` on a fresh clone should work, right? I get timeout errors in test_closing. What is the best way to reach out for help in such a case?

**Comment by cdecker** [comment]
> Thanks for taking the time to respond with words of encouragement and a willingness to help (and I also appreciate the existing documentation insofar as it goes). Git got clearer to me today. The biggest problem I am still struggling with at this point is running the tests, which of course is critical to any development. For example, `make pytest` on a fresh clone should work, right? I get timeout errors in test_closing. What is the best way to reach out for help in such a case?

No problem, it's though to start on a new project, especially when others have been working on it for years, and have become blind to the issues new contributors face.

Happy to help you here, or on IRC. It might just be a flaky test tbh, the test suite has quite a few cornercases, and configuration options, that may be stomping on each others toes (and I take full responsibility for this monstrosity), so if you share the error, we can try and figure it out together.

**Comment by rl-d** [comment]
I really appreciate you taking the time for me, and I will be staying online today and tomorrow in case you are available to help me get through this.

For the sake of not overlooking any possible newbie mistakes, here's how I'm working: Ubuntu 20.04, relatively fresh install on a VM dedicated to c-lighting dev. I went through an introductory vid by Rusty to get started with building and running start_ln for manual testing--that went pretty much flawlessly. And as you can see I had no trouble building and manually testing the code in this PR. So that much was virtually problem free and a joy.

Now for the sake of eliminating possible problems introduced by myself, I moved my whole lightning dir (to "lightning-firstissue") to be able to clone again and start with a clean slate. I've rm -rf'd this a couple of times to re-try different variations, and it goes something like this:

~/src$ ls
ccan  lightning-firstissue  lightning-rfc
~/src$ git clone https://github.com/ElementsProject/lightning.git
Cloning into 'lightning'...
remote: Enumerating objects: 79517, done.
remote: Counting objects: 100% (1633/1633), done.
remote: Compressing objects: 100% (774/774), done.
remote: Total 79517 (delta 921), reused 1373 (delta 823), pack-reused 77884
Receiving objects: 100% (79517/79517), 42.87 MiB | 611.00 KiB/s, done.
Resolving deltas: 100% (60304/60304), done.
~/src$ cd lightning
~/src/lightning$ ./configure --enable-developer
Compiling ccan/tools/configurator/configurator...done
...
~/src/lightning$ make
...
~/src/lightning$ sudo make install
[sudo] password for c-lightning:
...
~/src/lightning$ make pytest
mv: replace 'tools/headerversions', overriding mode 0755 (rwxr-xr-x)? y
PYTHONPATH=/home/c-lightning/src/lightning/contrib/pyln-client:/home/c-lightning/src/lightning/contrib/pyln-testing:/home/c-lightning/src/lightning/contrib/pyln-proto/:/home/c-lightning/src/lightning/external/lnprototest:/home/c-lightning/src/lightning/contrib/pyln-spec/bolt1:/home/c-lightning/src/lightning/contrib/pyln-spec/bolt2:/home/c-lightning/src/lightning/contrib/pyln-spec/bolt4:/home/c-lightning/src/lightning/contrib/pyln-spec/bolt7 TEST_DEBUG=1 DEVELOPER=1 VALGRIND=1 python3 -m pytest tests/ -v -p no:logging  -x
======================================= test session starts ========================================
platform linux -- Python 3.8.10, pytest-6.1.2, py-1.10.0, pluggy-0.13.1 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/c-lightning/src/lightning
plugins: timeout-1.4.2, rerunfailures-9.1.1, flaky-3.7.0, forked-1.3.0, xdist-2.2.1
collected 465 items

tests/test_closing.py::test_closing
...
In this case, I had a couple of tests succeed, unlike yesterday, but it still failed during test_closing_torture:

===================================== short test summary info ======================================
FAILED tests/test_closing.py::test_closing_torture - TimeoutError: Unable to find "[re.compile('S...
ERROR tests/test_closing.py::test_closing_torture - ValueError:
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 2 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
========================= 1 failed, 4 passed, 1 error in 936.46s (0:15:36) =========================

The exception was:

ERROR:BitcoindProxy:Exception on / [POST]
Traceback (most recent call last):
...
ConnectionRefusedError: [Errno 111] Connection refused

Could not connect to bitcoind using bitcoin-cli. Is bitcoind running?

Make sure you have bitcoind running and that bitcoin-cli is able to connect to bitcoind.

You can verify that your Bitcoin Core installation is ready for use by running:

    $ bitcoin-cli -regtest -datadir=/tmp/ltests-sya9hten/test_closing_torture_1/lightning-6/ -rpcport=37675 -rpcuser=... -rpcpassword=... echo 'hello world'

lightningd-6: 2021-09-16T09:59:03.157Z **BROKEN** plugin-bcli: \nCould not connect to bitcoind using bitcoin-cli. Is bitcoind running?\n\nMake sure you have bitcoind running and that bitcoin-cli is able to connect to bitcoind.\n\nYou can verify that your Bitcoin Core installation is ready for use by running:\n\n    $ bitcoin-cli -regtest -datadir=/tmp/ltests-sya9hten/test_closing_torture_1/lightning-6/ -rpcport=37675 -rpcuser=... -rpcpassword=... echo 'hello world'\n
lightningd-6: 2021-09-16T09:59:03.221Z INFO    plugin-bcli: Killing plugin: exited before we sent init
The Bitcoin backend died.
make: *** [Makefile:414: pytest] Error 1
~/src/lightning$

It's probably just something I did wrong... but on my own I'm kinda stuck not knowing what to look for. (I will try to get IRC set up...)

**Comment by vincenzopalazzo** [comment]
Hi @rl-d,

Do you have bitcoin core installed in your system? From the following error, it looks like it is missed

```
Is bitcoind running?
Make sure you have bitcoind running and that bitcoin-cli is able to connect to bitcoind.\n\nYou can verify that your Bitcoin Core installation is ready for use by running:\n\n $ bitcoin-cli -regtest -datadir=/tmp/ltests-sya9hten/test_closing_torture_1/lightning-6/ -rpcport=37675 -rpcuser=... -rpcpassword=... echo 'hello world'\n
lightningd-6: 2021-09-16T09:59:03.221Z INFO plugin-bcli: Killing plugin: exited before we sent init
The Bitcoin backend died.
```

> (I will try to get IRC set up...)

If you are find something easy to use, you can try to use [element](https://element.io/), with the matrix bridge, here there is a small guide on how to use the brige https://gist.github.com/LorbusChris/d44a5dd4671796cbf8528ca133dd909b

**Comment by rl-d** [comment]
Thanks, bitcoind is surely the problem... I knew it must be something simple...and I should have thought of that based on the error message.

**Comment by rl-d** [comment]
Ok, so this opens a bunch of questions for me. What exactly are the tests expecting in the way of bitcoind? Should it be an operational/synced node (which obviously I don't have on this box)? Should I start it with any special parameters? I feel like I'm a little in the dark as far as what's expected here.

**Comment by vincenzopalazzo** [comment]
>Ok, so this opens a bunch of questions for me. What exactly are the tests expecting in the way of bitcoind? Should it be an operational/synced node (which obviously I don't have on this box)? 

Mh, a lightning network node works with a bitcoin backend under the hood, so if you want to run a lightning node you must have a bitcoind in the system (or remotely). It is not used in the test, if you try to run also an ln node in production, you need a bitcoind backend.

>Should I start it with any special parameters? I feel like I'm a little in the dark as far as what's expected here.

For the test, it is c-lightning that takes care of how to set up the env. for production, you need to play around depending on your local config.

**Comment by rl-d** [comment]
You know what? I think I'm in the wrong profession. Or the wrong reality... Or the wrong era. Or maybe all of the above.

I see I have made way too many assumptions on my part and tried to race into this too fast. Too fast for my own mind to adapt. I thought I was working in the sandboxy world of simulated nodes. You mean this isn't a dream? You mean I'm really standing here in public with nothing but my underwear on?

I'm definitely going to have to admit that I'm not prepared to finish what I started here, because my first dev bitcoin node is still in the process of syncing (via sat) and that won't be done for a while.

I'm really sorry for wasting both of your time and I will go back to my crayons now.

**Comment by vincenzopalazzo** [comment]
>I'm definitely going to have to admit that I'm not prepared to finish what I started here, because my first dev bitcoin node is still in the process of syncing (via sat) and that won't be done for a while.

You don't need to sync a real to run with the c-lightning test, you need only to have a bitcoind installed. After that, if you run the command make pytest, you no longer see the previous error that you posted. All the config work it is made from c-lightning under the hood in a regtest network, that is a fake bitcoin network

If you want to run a bitcoin node in an easy way you can follow [this guide here ](https://vincenzopalazzo.medium.com/run-c-lightning-in-a-light-a-resilient-way-90917585859b)

To cite @cdecker, so take your time if you need more time, take it.
> it's though to start on a new project, especially when others have been working on it for years, and have become blind to the issues new contributors face.

**Comment by rl-d** [comment]
> > I'm definitely going to have to admit that I'm not prepared to finish what I started here, because my first dev bitcoin node is still in the process of syncing (via sat) and that won't be done for a while.
> 
> You don't need to sync a real to run with the c-lightning test, you need only to have a bitcoind installed. After that, if you run the command make pytest, you no longer see the previous error that you posted. All the config work it is made from c-lightning under the hood in a regtest network, that is a fake bitcoin network

Okay, this is the belief I was operating on, and therefore I am baffled why the tests are failing, because bitcoind IS installed. Then I must just be missing a command to make the magic work, but what?

I'm here dedicated to getting through this. I don't want to take more time waiting before getting involved if I can avoid it, but I have to be honest that I don't have a real node to play with yet if that is a requirement to successfully develop on c-lightning.



> 
> If you want to run a bitcoin node in an easy way you can follow [this guide here ](https://vincenzopalazzo.medium.com/run-c-lightning-in-a-light-a-resilient-way-90917585859b)
> 
> To cite @cdecker, so take your time if you need more time, take it.
> 
> > it's though to start on a new project, especially when others have been working on it for years, and have become blind to the issues new contributors face.

**Comment by rl-d** [comment]
Status log. The tests could finally be run successfully with VALGRIND=0 on a fresh clone, without changes. Seeing all the tests run successfully was an eye opener on a couple of counts: 1. seeing the time it takes to run the tests and 2. how many instances of lightningd and bcli are involved throughout the testing has given me an appreciation for the breadth and depth of the tests. It is clearer why folks don't necessarily run them as a whole, and it has tempered my ambition to run them too often as well. But it is incredibly satisfying to have such robust testing being performed and for the developer to be able to run them locally when needed.

As a newcomer, I'll mention that I think maybe it would be good for us to document how to run these types of tests in the Testing section of https://lightning.readthedocs.io/HACKING.html. This is one area where I felt the documentation was lacking, despite being excellent overall. In the docs overall, a good balance is struck between including enough information and not being too long and tedious. In this case, this is something pretty valuable that I think should really be there. I'll put that on my personal to-do list to add a line for `make pytest` and a brief explanation of what to expect.

However, I'm not entirely confident that everything else in that section shouldn't also be looked over by someone who knows how it should work, because frankly I had trouble getting anything in that section to work as one would expect. (In fact, most of the special build targets are still a mystery to me...this is the kind of thing I would like to see in a developer's orientation video or something like that...an overview of all the code generators and other project-uniuqe things that developers will likely need to touch at some point) Perhaps all this could be chalked up to newbie growing pains, but at the very least, running a successful test suite is such an important positive confirmation to a newcomer that it would be worth making sure they know how to get that confirmation to be sure all is well. It's the "Hello, world!" of building your own developer copy of a repo. I'm glad to at least have had the manual start_ln demo up front, which did provide some of that for me.

Alright, this wasn't meant to suggest more work for a busy team, but since there was mention of being sensitive to the experience of newcomers, at least it's documented here and maybe even I can come back to these topics myself someday when I've gotten some of these things under my belt enough to help out in that area. I'm more than happy to "help around the house" everywhere I can while I'm still getting up to speed on the real stuff that makes (and will make) lightning what it is (and will be).

It was a hard day. I made mistakes. I said things and reacted in ways I wish I didn't. It was also a very rewarding day. I don't feel worthy of rubbing shoulders with you all, but I hope to contribute something worth your investment in helping me get up to speed. Sitting down and focusing on this today has been productive even if progress wasn't on the actual subject matter of the PR.

**Comment by vincenzopalazzo** [comment]
Happy that you finally got what is the problem with your env. However, let me say some things about the doc.

>As a newcomer, I'll mention that I think maybe it would be good for us to document how to run these types of tests in the Testing section of https://lightning.readthedocs.io/HACKING.html. This is one area where I felt the documentation was lacking, despite being excellent overall. In the docs overall, a good balance is struck between including enough information and not being too long and tedious. In this case, this is something pretty valuable that I think should really be there.

In the doc, it is documented that valigrid need to be disabled 
https://lightning.readthedocs.io/HACKING.html#testing in fact all the commands are with the `VALGRIND=0`, and the `make pytest` it is a hack made by @cdecker (I think it is a hack, right?) because at the moment the command `make check` has a problem during the check of the main page https://github.com/ElementsProject/lightning/pull/4782. So I don't think that the command `make pytest` needs to have a place in the doc IMO.

>It was a hard day. I made mistakes. I said things and reacted in ways I wish I didn't. 

I personally start with this way in all the new projects :-D, it is a too easy start in some projects that you know all about, but a good way to learn new stuff is sitting in front of the desk and experiment. I started in this way one year ago, and for me C and make was scary at the time, node there is nothing more excited (except the regex :-P)

>I would like to see in a developer's orientation video or something like that

This is something that I start to think about some week ago, maybe I can share some my dev day on c-lightning and related projects it is not a bad idea. There is @niftynei that make some [dev video](https://www.twitch.tv/niftynei) but it is a protocol dev :) stuff too complex sometimes. Maybe a c-lightning 101 will be cool.

**Comment by rl-d** [comment]
I want to correct some misstatements in my previous comment and itemize exactly what surprises me as a newbie in the Testing section of https://lightning.readthedocs.io/HACKING.html. (The following errors are seen on a fresh clean copy of the repo with no additional changes.)

`make check` - I still haven't figured out what I'm doing wrong here. It results in:
fetchinvoice-noconnect documented but not in schema!
make: *** [doc/Makefile:152: check-config-docs] Error 1

`make full-check` - Same as above.

`make check-source` - This actually worked for me, once I added the rfc repo and tweaked permissions on tools/lightning-hsmtool, so that was great to be able to run that and see some green output. Very nice.

`make update-mocks` - I actually needed this at some point but it wouldn't run to completion for me, so I ended up hacking in my stub manually. Still not sure what I'm doing wrong here. The error happens on:
mocking out common/test/run-route.c:
cc: error: wire/peer_exp_wiregen.c: No such file or directory
cc: fatal error: no input files
compilation terminated.
make[1]: *** [Makefile:278: wire/peer_exp_wiregen.o] Error 1
make: *** [Makefile:633: update-mocks/common/test/run-route.c] Error 1

`PYTHONPATH=... py.test -v tests/` - This is the big one I've been wrestling with. It's here in the doc after all, albeit in a different form from `make pytest`, and that's fine... I had overlooked that this command was running these same tests. (It pays to "readthedocs"...I recognize I need to re-read them more.) So, nothing more to say on that...that's great that its included in the doc unlike I had said in my previous comment...sorry for that.

`make check-python` - This also works for me now.

So in summary, I am running into two errors out of the box (not counting the black-box tests). It's my desire to understand what I'm doing wrong in those cases.

**Comment by vincenzopalazzo** [comment]
>make check - I still haven't figured out what I'm doing wrong here. It results in:
fetchinvoice-noconnect documented but not in schema!
make: *** [doc/Makefile:152: check-config-docs] Error 1

>make full-check - Same as above.

Answer covered inside this PR https://github.com/ElementsProject/lightning/pull/4782 and described in the following issue https://github.com/ElementsProject/lightning/issues/4773

**Comment by rl-d** [comment]
> Happy that you finally got what is the problem with your env. However, let me say some things about the doc.
> 
> > As a newcomer, I'll mention that I think maybe it would be good for us to document how to run these types of tests in the Testing section of https://lightning.readthedocs.io/HACKING.html. This is one area where I felt the documentation was lacking, despite being excellent overall. In the docs overall, a good balance is struck between including enough information and not being too long and tedious. In this case, this is something pretty valuable that I think should really be there.
> 
> In the doc, it is documented that valigrid need to be disabled
> https://lightning.readthedocs.io/HACKING.html#testing in fact all the commands are with the `VALGRIND=0`, and the `make pytest` it is a hack made by @cdecker (I think it is a hack, right?) because at the moment the command `make check` has a problem during the check of the main page #4782. So I don't think that the command `make pytest` needs to have a place in the doc IMO.

Aha, (sorry I posted before seeing your reply) thanks for clarifying! Without knowing there was a known problem with `make check` I could only assume I was doing something wrong or the instructions were out of date or whatever, but I'm glad to know "It's not just me". (And yes, I need to re-read the docs more).

> > It was a hard day. I made mistakes. I said things and reacted in ways I wish I didn't.
> 
> I personally start with this way in all the new projects :-D, it is a too easy start in some projects that you know all about, but a good way to learn new stuff is sitting in front of the desk and experiment. I started in this way one year ago, and for me C and make was scary at the time, node there is nothing more excited (except the regex :-P)

Yes, I see I'm a bit rusty in some areas (not to mention new on the project) and I'm noticing I've also gotten more impatient over the years and want to understand things faster than maybe is good.

> > I would like to see in a developer's orientation video or something like that
> 
> This is something that I start to think about some week ago, maybe I can share some my dev day on c-lightning and related projects it is not a bad idea. There is @niftynei that make some [dev video](https://www.twitch.tv/niftynei) but it is a protocol dev :) stuff too complex sometimes. Maybe a c-lightning 101 will be cool.

Something like that... on the other hand, there are many vids out there I haven't seen yet, so maybe this is an area where I need to take it easy and keep learning from what's out there and not always ask for more things of the other developers who are already doing a lot to help us get on board. Thanks for your reply...

**Comment by rl-d** [comment]
Status:
- Test cases have been written and tested in my local repo.
- The deprecation of the channels from listpeers has caused some collateral damage to the topology plugin which I plan to investigate before finalizing.
